SDR-90 특정 음식점을 상세 조회할 수 있는 API 추가

jinan159 opened the pull request:
### ❗️ 이슈 번호

<strong>
Closes #8
</strong>

<br><br>

### 📝 구현 내용

- [x] 음식점 도메인 정의
- [x] 음식점 조회 API 구현
  - 인수테스트 작성
  - 서비스 통합테스트 작성
  - 도메인 단위테스트 작성
  - API 문서 작성

<br><br>

### 🙇🏻‍♂️ 리뷰어에게 부탁합니다!

1. `camel case` 와 `snake case` 중 택 1 하여, 통일 필요
  - `ReviewInsertRequest` : `camel case`
  - `OAuthTokenResponse` : `snake case`
  -> 일단 `StoreFindResponse` 는 `camel case` 로 작성하였음

2. Response DTO 에 record 적용 제안
  - record 클래스를 적용하면 Immutable class 를 손쉽게 만들 수 있는 장점이 있음
    - 소스코드가 깔끔해지는 장점이 있음
  - Jackson 라이브러리가 record 클래스의 직렬화/역직렬화를 지원함
  - 각 필드에 Annotation 도 붙일 수 있어서, 커스텀 가능
    - Validation, Json 관련 Annotation 등
  - ⚠ 주의할 점
    - `equals`, `hashcode`, `toString` 가 묵시적으로 override 되는것을 인지해야함
    - 라이브러리가 `getter`, `setter` 를 반드시 필요로 하는 경우 호환이 안될 수 있음
      - 사례 : [이 글](https://techblog.gccompany.co.kr/우리팀이-jdk-17을-도입한-이유-ced2b754cd7) 에서 `Record Data Class 추가` 부분 


<br><br>

### 💡 참고 자료

- [Oracle docs - Record](https://docs.oracle.com/en/java/javase/15/language/records.html)

jinan159 committed:
feat: :sparkles: StoreName 도메인 클래스 및 테스트 코드 추가

jinan159 committed:
feat: :sparkles: ContactNumber 도메인 클래스 및 테스트 코드 추가

jinan159 committed:
feat: :sparkles: Address 도메인 클래스 및 테스트 코드 추가

jinan159 committed:
feat: :sparkles: StoreLocation 도메인 클래스 및 테스트 코드 추가

jinan159 committed:
feat: :sparkles: Store 도메인 정의

- StoreName 추가
- ContactNumber 추가
- Address 추가
- StoreLocation 추가

jinan159 committed:
feat: :sparkles: StoreService 에 특정 키워드가 포함된 가게 목록 조회 기능 및 테스트 코드 추가

- StoreService 추가
- StoreRepository 추가
- StoreFindIntegrationTest 추가

jinan159 committed:
refactor: :construction_worker: StoreService 에서 가게 검색 결과로 DTO 를 반환하도록 기능 및 테스트 변경

- Store : Embedded 객체의 값을 넘겨주는 getter 메소드 추가
- StoreFindIntegrationTest : 가게 검색 서비스 통합 테스트 변경
- StoreFindResponse : 가게 검색 결과를 담는 DTO 추가
- StoreService : StoreFindResponse 를 반환하도록 변경

jinan159 committed:
feat: :sparkles: 가게 조회 API 및 인수테스트 추가

jinan159 committed:
test: :white_check_mark: 가게 조회 API 인수테스트 코드 변경

- DatabaseConfigurator : 가게 정보 추가
- StoreAcceptanceTest : DatabaseConfigurator 에서 초기화된 Test fixture 를 사용하도록 변경
- StoreSnippet : API 관련 Snippet 을 따로 관리
- InitAcceptanceTest : 불필요한 코드 제거

jinan159 committed:
docs: :memo: 음식점 조회 API 문서 작성

jinan159 commented on be/src/main/java/com/jjikmuk/sikdorak/store/controller/response/StoreFindResponse.java:
```diff
@@ -0,0 +1,24 @@
+
+import com.jjikmuk.sikdorak.store.domain.Store;
+
+public record StoreFindResponse(long id,
```
이 클래스에 record 클래스를 적용해보았습니다.
다른분들의 의견이 궁금해요!

seokho-ham replied:
p3; 확실히 코드 자체가 간결해지는 장점이 있군요!

lombok 어노테이션을 사용하지 않아도 `getter`와 `requiredArgsConstructor`를 만들어준다니..
(`getter`가 필드명으로 생성되는점은 아직 익숙하지 않아서 오히려 헷갈릴 수 있긴 할거 같다는 생각도 듭니다ㅎㅎ)

java 17을 사용하기로 한 이유 중 한가지가 record 사용의 이유였던거 같아서 전 이번기회에 써보는것도 좋을거 같습니다!

seokho-ham commented:
OAuthTokenResponse는 로그인 과정에서 카카오로부터 받는 액세스 토큰을 담는 dto를 의미합니다! 
네이밍 전략을 저렇게 설정해주지 않으면 역직렬화가 되지 않아서 설정했어요~

즉, 저희 클라이언트에게 제공하는 데이터 형식은 `camel case`로 가는게 맞다고 생각합니다~
(dto 이름 때문에 혼동을 줬군요😭)

seokho-ham commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.regex.Pattern;
+
+@Embeddable
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
p3; AccessLevel 설정 Good!! 👍

seokho-ham commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private boolean isContainsOnlyKoreanAndNumber(String text) {
+        return Pattern.matches("^[가-힣 0-9]*$", text);
```
p2; 정규식을 상수로 두는건 어떨까요?

jinan159 replied:
그게 괜찮겠네요!
의미를 전달할 수 있는 이름으로 명명해서 상수로 빼놓겠습니닷

seokho-ham commented on be/src/test/java/com/jjikmuk/sikdorak/store/domain/StoreLocationTest.java:
```diff
@@ -0,0 +1,58 @@
+        class Context_with_invalid_storePoint {
+
+            @ParameterizedTest
+            @CsvSource({
```
p3; null이나 빈값에 대한 케이스도 추가되면 어떨까요?

jinan159 replied:
StoreLocation 객체에 null 검증로직은 있는데, 테스트 케이스에서 검증을 못하고 있었군요!
이 케이스도 추가하는게 더 좋을 것 같네요 ㅎㅎ👍👍

ku-kim commented on be/src/docs/asciidoc/index.adoc:
```diff
@@ -14,3 +14,7 @@ notification-api-docs
 == 이슈 API
 
 include::{docdir}/review/review.adoc[]
+
+== 음식점 API
+
+include::{docdir}/store/store.adoc[]
```
Good!

ku-kim commented on be/src/docs/asciidoc/store/store.adoc:
```diff
@@ -0,0 +1,13 @@
+=== 이름으로 음식점 조회
+
+API : `GET /api/stores`
+
+=== `200 OK`
+
+==== `Request`
+
+operation::store_acceptance_test/find_store_by_name_containing_success[snippets='http-request,request-parameters']
+
+==== `Response`
+
+operation::store_acceptance_test/find_store_by_name_containing_success[snippets='http-response,response-fields']
```
테스트코드 포맷 지켜줬군요! 👍

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/controller/StoreController.java:
```diff
@@ -0,0 +1,29 @@
+    private final StoreService storeService;
+
+    @GetMapping("api/stores")
+    public CommonResponseEntity<List<StoreFindResponse>> findStore(@RequestParam("storeName") String storeName ) {
```
~~RequestBody를 storeName으로 지정한 이유가 있을까요?
혹시 스토어 이름이 유니크한가 고민이네요.
예를 들어 `할머니 순대국` storeName이 있다고 한다면 중복된 경우가 많지 않을까요.. ㅜ~~

p2;  findStore() 보다 자세한 네이밍이 있으면 좋을 거 같아요!
e.g. `findStoreName` , `readStoreName`, `readLikeStoreName`

jinan159 replied:
컨트롤러의 메소드는 호출하지 않는다 생각해서 그랬던 것 같습니닷😅
자세한 네이밍으로 변경할게요!

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -0,0 +1,60 @@
+public class Address {
+
+    private static final int BASEADDRESS_LENGTH = 255;
+    private static final int DETAILADDRESS_LENGTH = 50;
```
DETAILADDRESS_LENGTH 상수가 사용되고 있지 않는 것 같은데 맞을까요?

jinan159 replied:
아 `detailAddress` 를 검증할때 `DETAILADDRESS_LENGTH` 를 사용해야하는데 `BASEADDRESS_LENGTH` 를 사용했네요;;
이런 실수를..
수정해놓겠습니닷

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Address.java:
```diff
@@ -0,0 +1,60 @@
+        // detailAddress 는 null 이 아닌 경우, not empty, 50자 제한, 한글 숫자만
+        if (Objects.nonNull(detailAddress) &&
+                (detailAddress.isBlank() ||
+                        detailAddress.length() > BASEADDRESS_LENGTH ||
```
```suggestion
                        detailAddress.length() > DETAILADDRESS_LENGTH ||
```

BASEADDRESS_LENGTH 가 아니라 DETAILADDRESS_LENGTH 기준이라고 생각했는데 다른 의도가 있는걸까요?
DETAILADDRESS_LENGTH 주소 테스트가 없는 것 같아요..!

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/ContactNumber.java:
```diff
@@ -0,0 +1,46 @@
+@Getter
+public class ContactNumber {
+
+    /**
+     * 허용되는 전화번호 형식 <br>
+     * - 00-000-0000 <br>
+     * - 00-0000-0000 <br>
+     * - 000-000-0000 <br>
+     * - 000-0000-0000
+     */
```
주석 좋아요 👍

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/Store.java:
```diff
@@ -1,21 +1,68 @@
 @Entity
-@Getter
-
+@NoArgsConstructor // for @Entity
```
// for `@Entity` 주석은 왜 남겨져있는거에요?

jinan159 replied:
예전에 코드리뷰 받을때 `@NoArgsConstructor` 왜 붙여놓은건지 여러번 질문을 받은적이 있었습니다.
그래서 `@Entity` 가 `@NoArgsConstructor` 가 필요로 한다는걸 명시하기 위해서 붙였어요!

ku-kim replied:
아하 그런 이유군요!!

ku-kim commented on be/src/main/java/com/jjikmuk/sikdorak/store/domain/StoreLocation.java:
```diff
@@ -0,0 +1,45 @@
+    private static final double LONGITUDE_MIN = -180.0;
+    private static final double LONGITUDE_MAX = 180.0;
+
+    private Point storePoint;
```
Location을 Point 객체로! Good

ku-kim commented on be/src/test/java/com/jjikmuk/sikdorak/acceptance/StoreAcceptanceTest.java:
```diff
@@ -0,0 +1,68 @@
+        Store savedStore = testData.store;
+
+        ResponseCodeAndMessages expectedCodeAndMessage = ResponseCodeAndMessages.STORE_FIND_SUCCESS;
+        String storeNameSearchKeywork = "가게";
```
Request 값과 savedStore 값이 별도로 구분되어 있어 testData.store가 변경되면 테스트가 깨질 거 같아요

jinan159 replied:
저장된 스토어의 값을 활용하는 방식을로 변경해보겠습니다!

ku-kim commented on be/src/test/java/com/jjikmuk/sikdorak/integration/store/StoreFindIntegrationTest.java:
```diff
@@ -0,0 +1,115 @@
+
+        Store savedStore = null;
+
+        @BeforeEach
+        void setUp() {
+            storeRepository.deleteAll();
+            savedStore = storeRepository.save(new Store(
+                    STORE_NAME,
+                    CONTACT_NUMBER,
+                    BASE_ADDRESS,
+                    DETAIL_ADDRESS,
+                    LATITUDE,
+                    LONGITUDE
+            ));
+        }
```
InitIntegrationTest 에 초기화 데이터가 있는데 savedStore으로 테스트한 이유가 있을까요?

jinan159 replied:
이 부분은 기존에 작성해놓았던 부분인데, 테스트코드 리펙토링을 하다 놓친 부분인 것 같네요!
수정할게요!

ku-kim commented on be/src/test/java/com/jjikmuk/sikdorak/integration/store/StoreFindIntegrationTest.java:
```diff
@@ -0,0 +1,115 @@
+            List<StoreFindResponse> stores = storeService.findStoresByStoreNameContaining(storeName);
+
+            // then
+            assertThat(stores).isNotNull();
+            assertThat(stores).isEmpty();
```
Q) List<> 에 isEmpty() 검증한다는 것은 리스트 사이즈가 0이라는 뜻일까요?

jinan159 replied:
넵넵 맞습니다!
리스트 길이가 0인지 확인하는 코드입니다.

ku-kim reviewed (commented):
고생 많으셨어요 Jay!

jinan159 committed:
refactor: :construction_worker: Address 검증로직 리펙토링

- detailAddress 길이 검증시, DETAILADDRESS_LENGTH 를 사용하도록 변경
- 텍스트에 한글숫자만 있는지 검증하는 메소드의 정규식을 상수로 추출

jinan159 committed:
test: :white_check_mark: StoreLocation 에 null 인 경우에 대한 테스트 케이스 추가

jinan159 committed:
refactor: :construction_worker: StoreController 음식점 조회 메소드 명칭 변경

jinan159 committed:
test: :white_check_mark: 가게 이름 검색 테스트 성공케이스에서, 저장된 가게 이름을 사용하도록 변경

ku-kim reviewed (approved):
수고하셨어요~!
